src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,405 @@
+    ensure_full_row_binlog_format(&mut conn).await?;
+    ensure_replication_commit_order(&mut conn).await?;
+
+    // To start the stream we need to provide a GTID set of the transactions that we've 'seen'
+    // and the server will send us all transactions that have been committed after that point.
+    // NOTE: The 'Gno' intervals in this transaction-set use an open set [start, end)
+    // interval, which is different than the closed-set [start, end] form returned by the
+    // @gtid_executed system variable. So the intervals we construct in this GTID set
+    // end with the value of the transaction-id that we want to start replication at,
+    // which happens to be our `resume_transaction_id`
+    // https://dev.mysql.com/doc/refman/8.0/en/replication-options-gtids.html#sysvar_gtid_executed
+    // https://dev.mysql.com/doc/dev/mysql-server/latest/classGtid__set.html#ab46da5ceeae0198b90f209b0a8be2a24
+    let seen_gtids = match resume_transaction_id.into() {
+        0 | 1 => {
+            // If we're starting from the beginning of the binlog or the first transaction id
+            // we haven't seen 'anything'
+            Vec::new()
+        }
+        ref a => {
+            // NOTE: Since we enforce replica_preserve_commit_order=ON we can start the interval at 1
+            // since we know that all transactions with a lower transaction id were monotonic
+            vec![Sid::new(*gtid_source_id.as_bytes()).with_interval(GnoInterval::new(1, *a))]
+        }
+    };
+
+    // Request that the stream provide us with a heartbeat message when no other messages have been sent
+    let heartbeat = Duration::from_secs(3);
+    conn.query_drop(format!(
+        "SET @master_heartbeat_period = {};",
+        heartbeat.as_nanos()
+    ))
+    .await?;
+
+    // Try to generate a deterministic server-id for identifying us as a replica on the upstream mysql server.
+    // The value does not actually matter since it's irrelevant for GTID-based replication and won't cause errors
+    // if it happens to be the same as another replica in the mysql cluster (based on testing),
+    // but by setting it to a constant value we can make it easier for users to identify Materialize connections
+    let server_id = match config.id {
+        GlobalId::System(id) => id,
+        GlobalId::User(id) => id,
+        GlobalId::Transient(id) => id,
+        _ => unreachable!(),
+    };
+    let server_id = match u32::try_from(server_id) {
+        Ok(id) if id + REPLICATION_SERVER_ID_OFFSET < u32::MAX => id + REPLICATION_SERVER_ID_OFFSET,
+        _ => REPLICATION_SERVER_ID_OFFSET,
+    };
+
+    let repl_stream = match conn
+        .get_binlog_stream(
+            BinlogStreamRequest::new(server_id)
+                .with_gtid()
+                .with_gtid_set(seen_gtids),
+        )
+        .await
+    {
+        Ok(stream) => stream,
+        Err(mysql_async::Error::Server(ref server_err)) if server_err.code == 1236 => {
```

Comment:
Looks like this code has a name (`ER_SOURCE_FATAL_ERROR_READING_BINLOG`), let's define a constant for it. Also, it says this code only exists in MySQL 8.0.33 and above. I don't know if that changes anything to our assumptions. https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html#error_er_source_fatal_error_reading_binlog

Reply:
Ah, no, it existed in 5.7 too, just with a different name https://dev.mysql.com/doc/mysql-errors/5.7/en/server-error-reference.html#error_er_master_fatal_error_reading_binlog

Reply:
thanks for looking those up, will define a constant and include those links